sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -999,14 +1022,7 @@ public void onClaimFailed(PositionT position) {}
         }
       }
     } finally {
```

Comment:
I don't think we need this finally block since the only way out is via the break statement.

Any exceptional return will lead to the bundle being failed and these DoFns being discarded.

Ditto for the other finally block.

Reply:
Is that guaranteed somewhere? I'm concerned that if things change so that the exception is caught that we've left this object in an invalid state.  We had a recent bug in state where the user was able to catch the exception and the invalid state was then used. I don't think that is possible here but is it possible that we add some generic exception handling (perhaps by configuration) and would rather clean up as we go here? Or is that unlikely in your opinion?

Reply:
Note that this finally structure is used throughout to reset after processing.  If you do think it should be done, should I update the other places in this file too?

Reply:
Went ahead and updated the methods I changed, but left other finally blocks to cut down on timezone delays if that is the path desired.

Reply:
I think you missed pushing out your latest version of this code.

In general your right that a DoFn could throw an exception which is caught by another DoFn further up the stack. This wouldn't apply to the SplittablDoFn since it is at the root of the stage for all runners so we are ok. There is also the problem were the SplittableDoFn itself throws an exception during processElement and it hasn't yet escaped the DoFn (whether that exception was caused during usage of the RestrictionTracker or something else). In this scenario we need to assume that the RestrictionTracker/WatermarkEstimatorState/DoFn state that getSize relies on isn't corrupted and a proper split can be returned. The only way I think we could resolve this completely is if we move the splitting logic to actually happen on the bundle processing thread by inserting points where we check if there is an outstanding split and process it but there is no time bound on how long we may have to wait before handling a split.

We can solve the DoFn exception propagation problem by ensuring that uncaught exceptions that escape a DoFn fail the bundle in the PCollectionConsumerRegistry and also plug this logic into places when the DoFn invokes system code that causes a terminal bundle failure (e.g. state gets corrupted). The issue comes down to whether we trust that the user is specifically handling the failure and choosing to continue or are they just blindly catching everything.

Reply:
Oops, yes I didn't push.

If I understand correctly, in this particular case we are guaranteed that if an exception escapes it will fail because it is the root. So leaving the code as is.  Let me know if there is something else you'd like changed in this PR.